### PR TITLE
Flash decoding round 1; AIMIGRAPHX-242

### PR DESCRIPTION
**Pull request by bdevorem**
First round in addressing #4334

## Motivation

rocMLIR has implemented split kv and GQA, which enables us to implement flash decoding. Now we need to add this to the migraphx side.

## Technical Details

This supports rewriting attention with a specified number of splits along the KV sequence length dimension, such that those chunks of the attention can be computed in parallel. Afterwards, the results are rescaled and combined. In practice, for migraphx, that rewriting looks like:

```
Q -> [B, M, k]
K -> [B, k, N]
V -> [B, N, D]

S = dot(Q, K)
P = softmax(S)
O = dot(P, V) # [B, M, D]
```
rewritten into 
```
Q -> [B, G, M, k] 
K -> [B, G, k, N/G]
V -> [B, G, N/G, D]

# first kernel
S = dot(Q, K)
P = softmax(S, axis=-1)
L = LSE(S) # [B, G, M, 1]
O' = dot(P, V) # [B, G, M, D]

# second kernel
m = max(L, axis=1) # [B, 1, M, 1],  global max for stability
alpha = exp(L - m) # [B, G, M, 1],  unnormalized weights
scale = alpha / sum(alpha, axis=1) # [B, G, M, 1],  normalized weights
R = mul(O', scale) # [B, G, M, D]
O = sum(R, axis=1) # [B, 1, M, D]
```

This PR also supports multi-headed attention, i.e. starting with Batch *and* Heads dimensions, instead of just Batch. 

I have added an env var `MIGRAPHX_ENABLE_FLASH_DECODING` where we can set explicitly the number of splits, e.g. `MIGRAPHX_ENABLE_FLASH_DECODING=2`, but also for testing, the splits can be directly set through the fuse_attention class.

Currently, this does not rewrite attention if it is LSE, nor if there is input fusion (which might not even happen at this stage, I've saved this problem for later). It derives the starting and rewritten shapes based on the gemm sizes, and then does the following:
1. inserts reshapes/broadcasts for Q, K, V
2. creates a new attention submodule with the new (reshaped) inputs
3. rebuilds the original attention submodule with updated reduction operators
4. inserts LSE at the end of the new submodule
5. returns both the partial out and the LSE
6. adds in the second kernel after unpacking the new submodule results
7. reshapes the final output back to what is expected

The logic for rebuilding the submodule could use some fine-tuning.

## Future work
- - [ ] if input fusion is allowed on attention, then need to handle this
- - [ ] if this needs to work for LSE attention, then need to handle this
- - [ ] split-kv case
- - [ ] add 4D test cases when rocmlir bug is resolved
- - [ ] make the 3D verify tests automatic
- - [ ] use fuse with special inserter for rebuilding submodule

## Changelog Category

- - [x] Added: New functionality.
- - [ ] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [ ] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.

### CODE REVIEW

**Review by anisha-amd — APPROVED** [review]

File: src/fuse_attention.cpp [hunk 1]
```diff
@@ -30,12 +30,28 @@
...
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
 namespace {
 
+MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_ENABLE_FLASH_DECODING);
+
+std::size_t get_num_splits()
+{
+    std::string value = string_value_of(MIGRAPHX_ENABLE_FLASH_DECODING{}, "0");
```
Comment (pfultz2) [inline src/fuse_attention.cpp]: The `value_of` function can be used to get an integer.

File: src/fuse_attention.cpp [hunk 2]
```diff
@@ -209,6 +225,352 @@ struct find_attention
...
+        {
+            auto param_ins               = submod->get_parameter(param_names[i]);
+            map_param_to_main[param_ins] = group_inputs[i];
+        }
+
+        // verify the mapping is correct
+        auto expected_inputs = submod->get_inputs(map_param_to_main);
+        assert(expected_inputs == group_inputs and "Mapped inputs don't match group inputs");
+        return map_param_to_main;
+    }
```
Comment (pfultz2) [inline src/fuse_attention.cpp:342]: I think you can use `get_ins_param_map` for this function.

File: src/fuse_attention.cpp [hunk 3]
```diff
@@ -209,6 +225,352 @@ struct find_attention
...
+    void rebuild_attention_submodule(
+        module& target_mod,
+        const module& source_mod,
+        const std::unordered_map<instruction_ref, instruction_ref>& param_map) const
+    {
+        // map from instructions in the old module to the new ones in the target module
+        std::unordered_map<instruction_ref, instruction_ref> map_old_to_new = param_map;
+        std::unordered_map<std::string, instruction_ref> softmax_parts;
+
+        for(auto it = source_mod.begin(); it != source_mod.end(); ++it)
```
Comment (pfultz2) [inline src/fuse_attention.cpp:353]: Use `insert_instructions` or `add_instructions` to copy the instructions over. It will take care of the mapping and traversal. There is an `inserter` function that can be used to change the operators as you insert them in.

File: src/fuse_attention.cpp [hunk 4]
```diff
@@ -209,6 +225,352 @@ struct find_attention
...
+            if(op.name() == "reduce_sum")
+                softmax_parts["sum_exp"] = new_ins;
+        }
+
+        // get the final partial output (O')
+        auto orig_return_ins        = std::prev(source_mod.end())->inputs().front();
+        auto partial_output_o_prime = map_old_to_new.at(orig_return_ins);
+
+        // calculate LSE = max(S) + log(sum(exp(S - max(S))))
+        assert(contains(softmax_parts, "max") and contains(softmax_parts, "sum_exp"));
```
Comment (pfultz2) [inline src/fuse_attention.cpp:420]: Just do a `find_if` for `reduce_max` and then insert the LSE operators. Then we can apply CSE to remove the duplicate calculations with softmax. This should be simpler then finding all the parts of softmax.

File: src/fuse_attention.cpp [hunk 5]
```diff
@@ -209,6 +225,352 @@ struct find_attention
...
+            m_flash_decode.add_parameter(v_name, shape{qkv_shapes[2].type(), vp_shape});
+
+        // build mapping for old params -> new params
+        std::unordered_map<instruction_ref, instruction_ref> map_old_params_to_new;
+        map_old_params_to_new[q_param] = new_q_param;
+        map_old_params_to_new[k_param] = new_k_param;
+        map_old_params_to_new[v_param] = new_v_param;
+
+        // don't simply fuse previous attn submod, need to rebuild all the ops
+        rebuild_attention_submodule(m_flash_decode, *submod, map_old_params_to_new);
```
Comment (pfultz2) [inline src/fuse_attention.cpp:537]: Actually you can use the `fuse` method here. The `inserter` function can be used to change the operators to handle the extra dimension, and the parameters will be inserted correctly based on the inputs.

File: src/include/migraphx/fuse_attention.hpp [hunk 6]
```diff
@@ -35,6 +37,11 @@ struct module_pass_manager;
 
 struct MIGRAPHX_EXPORT fuse_attention
 {
+    std::optional<std::size_t> flash_decoding_num_splits = std::nullopt;
+
+    fuse_attention() = default;
+    explicit fuse_attention(std::size_t num_splits) : flash_decoding_num_splits(num_splits) {}
```
Comment (pfultz2) [inline src/include/migraphx/fuse_attention.hpp]: If its possible, I would remove the constructor here, so we can use the name to initialize as it can be clearer.

File: test/fuse_attention.cpp [hunk 7]
```diff
@@ -38,11 +38,19 @@
...
 #include <utility>
 
+MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_ENABLE_FLASH_DECODING);
+
 static void run_pass(migraphx::program& p)
 {
     migraphx::run_passes(p, {migraphx::fuse_attention{}, migraphx::dead_code_elimination{}});
 }
 
+static void run_pass(migraphx::program& p, std::size_t num_splits)
```
Comment (pfultz2) [inline test/fuse_attention.cpp]: No need for an extra overload. I would just add the pass as a default parameter:

```cpp
static void run_pass(migraphx::program& p, migraphx::fuse_attention fa = {})
{
    migraphx::run_passes(p, {fa, migraphx::dead_code_elimination{}});
}
```

And then you can run this with `run_pass(p1, {.flash_decoding_num_splits=2})`.

File: src/fuse_attention.cpp [hunk 8]
```diff
@@ -209,6 +225,352 @@ struct find_attention
...
+                const auto& new_input_shape = new_inputs.front()->get_shape();
+                assert(original_axes.front() ==
+                           static_cast<int64_t>(ins->inputs().front()->get_shape().lens().size() -
+                                                1) or
+                       original_axes.front() == -1);
+                op.from_value(
+                    {{"axes", {static_cast<int64_t>(new_input_shape.lens().size() - 1)}}});
+            }
+            // TODO make less reliant on ops around it
+            else if(op.name() == "multibroadcast")
```
Comment (pfultz2) [inline src/fuse_attention.cpp:387]: We know which axis we inserted the groups so we can just insert that into the `out_lens` and that should work most of the time. For more robust solution is to either rewrite this to use 2 args version and then remove it afterwards or add an optional parameter to the inserter for the old_ins so we can find the second argument on the fly.

File: src/fuse_attention.cpp (hunk 5 above)
Comment (bdevorem) [reply src/fuse_attention.cpp:537]: ah that's much better, thanks will do

Comment (bdevorem) [reply src/fuse_attention.cpp:537]: I'll make this change in round 2

File: docs/reference/MIGraphX-dev-env-vars.rst [hunk 9]
```diff
@@ -160,6 +160,14 @@ Model performance tunable variables change the compilation behavior of a model.
 
       | Default: Split-k performance configurations are turned off.
 
+  * - | ``MIGRAPHX_ENABLE_FLASH_DECODING``
+      | Turns on flash decoding for attention fusion and sets the number of splits along the key-value sequence dimension.
+    
+    - | ``0``: Flash decoding is turned off (i.e., number of splits is 0).
+      | ``N`` (where N > 1): Enables flash decoding with N splits along the key-value sequence dimension. For example, ``2`` enables flash decoding with 2 splits, ``4`` with 4 splits, etc.
```
Comment (shivadbhavsar) [inline docs/reference/MIGraphX-dev-env-vars.rst:167]: Id call this flash_decoding_num_splits or something. "Enable" makes it seem like a on/off flag only

File: src/fuse_attention.cpp [hunk 10]
```diff
@@ -209,6 +214,395 @@ struct find_attention
...
+    }
+
+    void apply(module_pass_manager& mpm, const match::matcher_result& r) const
+    {
+        auto& mm            = mpm.get_module();
+        auto attn_group_ins = r.instructions["group"];
+        auto* submod        = attn_group_ins->module_inputs().front();
+
+        // TODO: for this pass of flash decoding, if LSE attn, do not do flash decoding
+        auto return_ins = std::prev(submod->end());
```
Comment (shivadbhavsar) [inline src/fuse_attention.cpp:435]: maybe add a quick assert here to make sure it last ins is really the return ins

File: src/fuse_attention.cpp [hunk 11]
```diff
@@ -209,6 +214,395 @@ struct find_attention
...
+        auto v_lens = input_shapes[2].lens();
+
+        // 3D: Q_lens = [B, M, k]
+        // 4D: Q_lens = [B, H, M, k]
+        size_t ndim = q_lens.size();
+        size_t n    = k_lens[ndim - 1];
+        size_t g    = groups;
+
+        // TODO: do we wanna support this differently?
+        assert(n % g == 0 and
```
Comment (shivadbhavsar) [inline src/fuse_attention.cpp:285]: might make sense to just not apply this flash decoding transformation in this case (ie. check this in the apply method and return). I assume you would need to add some sort of padding logic to support this case?

Comment (bdevorem) [reply src/fuse_attention.cpp:285]: I can change this PR to do that for sure. In the next round, when I add logic for deciding G, I was going to handle the padding

File: src/fuse_attention.cpp [hunk 12]
```diff
@@ -209,6 +214,400 @@ struct find_attention
...
+        auto k_lens = input_shapes[1].lens();
+        auto v_lens = input_shapes[2].lens();
+
+        // 3D: Q_lens = [B, M, k]
+        // 4D: Q_lens = [B, H, M, k]
+        size_t ndim = q_lens.size();
+        size_t n    = k_lens[ndim - 1];
+        size_t g    = groups;
+
+        // TODO: handle uneven splits; this is caught in `apply` for now
```
Comment (TedThemistokleous) [inline src/fuse_attention.cpp:284]: Make this an issue we can track

**Review by TedThemistokleous — APPROVED** [review]
LGTM.

Just fix CI
Make sure the TODOs are made into issues so you can tackle them/edge cases once this goes in. Good things to track as improvements.

**Review by anisha-amd — APPROVED** [review]
